### ISSUE TITLE
Automatic transaction classification using machine learning

### ISSUE THREAD

**Issue by nschoonbroodt** [nschoonbroodt/rusty-assets#31]
## Overview

Implement automatic transaction categorization that learns from existing classified transactions to suggest categories for new imports.

## Goals

- **Reduce manual work**: Auto-suggest categories based on transaction descriptions
- **Learn from patterns**: Use existing transaction history as training data
- **Improve over time**: Classification accuracy increases with more data
- **User control**: Allow manual override and feedback to improve suggestions

### Core Classification Engine
- **Feature extraction** from transaction descriptions (keywords, amounts, merchants)
- **Pattern recognition** for recurring transactions (salary, rent, subscriptions)
- **Similarity matching** against historical transactions
- **Confidence scoring** for classification suggestions

#### Option 1: Simple Rule-Based + Fuzzy Matching
```rust
// Fast to implement, no external ML dependencies
if description.contains("SALARY") || description.contains("PAYROLL") {
    suggest_category("Income::Employment", confidence: 0.95)
}

// Fuzzy string matching against known merchants
let similarity = jaro_winkler("AMAZON.COM", transaction.description);
if similarity > 0.8 {
    suggest_category("Expenses::Shopping::Online", confidence: 0.8)
}
```

#### Option 2: Traditional ML (scikit-learn style)
- **TF-IDF vectorization** of transaction descriptions
- **Naive Bayes** or **Random Forest** classifier
- Use `candle-core` or `smartcore` crates for Rust implementation
- Export training data to Python for model development, then port to Rust

#### Option 3: Lightweight Neural Networks
- **Word embeddings** for transaction descriptions
- **Simple feed-forward network** for classification
- Use `tch` (PyTorch bindings) or `candle-nn` for implementation

### Training Data Structure
```rust
pub struct TrainingExample {
    pub description: String,
    pub amount: Decimal,
    pub account_type: AccountType,
    pub category: Option<String>,
    pub merchant: Option<String>,
    pub user_confirmed: bool, // Feedback loop
}
```

### User Experience
- **Suggestion UI**: Show top 3 category suggestions with confidence scores
- **One-click acceptance**: Quick approval of suggestions
- **Feedback loop**: Learn from corrections and manual classifications
- **Batch classification**: Process multiple similar transactions at once

### Phase 1: Data Foundation
- [ ] Extract existing transaction patterns from database
- [ ] Build training dataset from historical transactions
- [ ] Implement basic text preprocessing (normalization, tokenization)

### Phase 2: Simple Classifier
- [ ] Rule-based classifier for obvious patterns (salary, utilities, rent)
- [ ] Fuzzy string matching for merchant recognition
- [ ] Basic confidence scoring system

### Phase 3: ML Enhancement
- [ ] Choose ML approach based on data size and performance requirements
- [ ] Implement feature extraction pipeline
- [ ] Train initial model on historical data
- [ ] Model evaluation and validation

### Phase 4: User Integration
- [ ] CLI commands for classification suggestions
- [ ] Batch import with auto-classification
- [ ] Feedback mechanism for model improvement
- [ ] Classification confidence thresholds

### Rust ML Ecosystem
- **candle-core**: Facebook's Candle framework for Rust
- **smartcore**: Pure Rust ML library
- **linfa**: Rust ML toolkit (scikit-learn equivalent)
- **tokenizers**: Hugging Face tokenizers for text processing

### Performance Requirements
- **Fast suggestions**: < 100ms for real-time classification
- **Memory efficient**: Work with limited training data
- **Incremental learning**: Update model with new classifications

### Privacy & Data
- **Local processing**: All ML runs locally, no data sent to external services
- **Data anonymization**: Option to exclude sensitive information from training
- **Export capability**: Allow users to export their classification models

## Success Criteria
- [ ] 80%+ accuracy on recurring transactions (salary, rent, utilities)
- [ ] 60%+ accuracy on merchant-based transactions
- [ ] Sub-100ms response time for classification suggestions
- [ ] User feedback integration improving accuracy over time

## Future Enhancements
- **Multi-language support**: Classification in different languages
- **Account-specific models**: Different classification patterns per account type
- **Collaborative filtering**: Optional sharing of anonymized classification patterns
- **Integration with bank APIs**: Real-time classification as transactions arrive

This feature would significantly reduce the manual effort in transaction categorization while keeping all processing local and private.
